packages/nitro-protocol/contracts/examples/EmbeddedApplication.sol

```diff
@@ -0,0 +1,438 @@
+contract EmbeddedApplication is
+    IForceMoveApp2 // We need a new interface to allow signedBy information into the validTransition function
+{
+    // 2-party SupportProof
+    struct SupportProof {
+        IForceMove.FixedPart fixedPart;
+        VariablePart[] variableParts; // either one or two states
+        uint48 turnNumTo;
+        IForceMove.Signature[2] sigs; // one for each participant
+        uint8[2] whoSignedWhat; // whoSignedWhat = [0,0] or [0,1] or [1,0]
+    }
+
+    // Finite states for J
+    enum AlreadyMoved {None, A, B, AB}
+
+    // Application Data for J
+    struct AppData {
+        bytes32 channelIdForX;
+        SupportProof supportProofForX;
+        AlreadyMoved alreadyMoved;
+    }
+
+    function validTransition(
+        VariablePart memory from,
+        VariablePart memory to,
+        uint48, // turnNumTo (unused)
+        uint256, // nParticipants (unused)
+        uint256, // signedByFrom (unused) - Who has signed the "from" state?
+        uint256 signedByTo // Who has signed the "to" state?
+    ) public override pure returns (bool) {
+        // parameter wrangling
+        bool signedByA = ForceMoveAppUtilities.isSignedBy(signedByTo, 0);
+        bool signedByB = ForceMoveAppUtilities.isSignedBy(signedByTo, 1);
+        AppData memory fromAppData = abi.decode(from.appData, (AppData));
+        AppData memory toAppData = abi.decode(to.appData, (AppData));
+        Outcome.AllocationItem[] memory fromAllocation = decode3PartyAllocation(from.outcome);
+        Outcome.AllocationItem[] memory toAllocation = decode3PartyAllocation(to.outcome);
+
+        require(
+            fromAllocation[0].destination == toAllocation[0].destination &&
+                fromAllocation[1].destination == toAllocation[1].destination &&
+                fromAllocation[2].destination == toAllocation[2].destination,
+            'destinations may not change'
+        );
+
+        require(fromAllocation[2].amount == toAllocation[2].amount, 'p2.amt !constant');
+        require(
+            fromAllocation[0].amount + fromAllocation[1].amount + fromAllocation[2].amount ==
+                toAllocation[0].amount + toAllocation[1].amount + toAllocation[2].amount,
+            'total allocation changed'
+        );
+
+        // Allowed named-state transitions are
+        //    AB
+        //    ^^
+        //   /  \
+        // A      B
+        // ^      ^
+        //  \    /
+        //    ABI
+
+        if (fromAppData.alreadyMoved == AlreadyMoved.ABI) {
+            require(
+                (toAppData.alreadyMoved == AlreadyMoved.A && signedByA) ||
+                    (toAppData.alreadyMoved == AlreadyMoved.B && signedByB),
+                'incorrect move from ABI'
+            );
+        } else {
+            // If a support proof has already been supplied, the current support proof must be greater
+            require(
+                toAppData.supportProofForX.turnNumTo > fromAppData.supportProofForX.turnNumTo,
```

Comment:
[pre-emptive edit]: The below comment probably requires #3556 to enable the proposed optimization.

--- 

There's probably an opportunity for optimization here: in the _beginning_ of the VFP, Alice & Bob don't need to sign preFS for `X`. They can sign preFS for `J`. Alice can make a state update in the channel once the VFP is done.

This is unsafe for Alice if Alice can _take money from Bob_ in her first move -- she could lose money that's rightfully hers, by not being allowed to to progress the channel. But it's fine for unidirectional Alice -> Bob channels.

💡 can we address that by _not storing a support proof_ in the app data of `J`? Or, not including `sigs` in `J.preFS`? (Alice & Bob will both sign `J.preFS`, so they're implicitly agreeing to the "preFS" that's embedded in `J.preFS.appData`.)

Reply:
This feels out-of-scope, to me. Perhaps this could go into a pitch for a future cycle?

Reply:
it is definitely out of scope for this PR! But, it plays into the analysis [here](https://www.notion.so/statechannels/Virtual-funding-optimizations-16998f6d3baa41a49dd69b9948721b66#db9744e2dfff4973b0b23d3de85d1f55), so I was hoping you had some thoughts around it already 😆

Reply:
Definitely sounds like an interesting idea: although ideally we don't want to rely on any particular property of the application rules of `X` (such as unidirectionality). 

I need to think more about refunding of guarantees. Also interesting, of course, but adding to the core nitro API throws up a lot of new things that can happen.